Ensure we don't settle with an outdated price

thomaseizinger opened the pull request:
Fixes #1057.

thomaseizinger committed:
Invert price-feed and projection dependencies

The price-feed is the more standalone component, hence it makes more
sense for the projection actor to depend on the price-feed rather than
the other way round.

This introduces a "polling" communication pattern but at the same time
simplifies the price-feed and will allow us to further embed it in the
actor system.

thomaseizinger committed:
Shorten field names

It only ever makes sense to store the address of an actor, not the actor
itself. Thus, a field named `xyz_actor` implies that it will be an
address. Furthermure, we can see from the type signature that it is an
address.

thomaseizinger committed:
Use whitespace to group related fields

thomaseizinger committed:
Embed price-feed actor in actor system

This allows us to access the latest quote from the price-feed via the
`LatestQuote` message directly within `propose_settlement`. This removes
the hack of converting the projected quote back into a
`bitmex_price_feed::Quote` within the taker's routes. It also increases
the encapsulation of our actor system and reduces setup boilerplate
within the binary.

To properly mock the price feed, we need to write the mock ourselves because
mockall doesn't allow us to _overwrite_ an expectation. That is, if we start
out with a mock of `None` for the current price, we cannot later override it
with `Some(quote)`. Writing the mock by hand allows us to do this and is not
actually much more code.

thomaseizinger committed:
Don't return 202 from endpoints

202 ACCEPTED specifies that a long running operation has been queued and
the result is not yet available. We are not doing that, all our
endpoints are synchronous in their execution.

Remove `status::Accepted` in favor of `()` which results in an empty
200 OK response.

thomaseizinger committed:
Send heartbeats as part of select loop

No need to spawn a separate task, we can just tick as part of the loop
within the EventStream.

thomaseizinger committed:
Ensure we don't settle with an outdated price

Fixes #1057.

luckysori commented on daemon/src/lib.rs:
```diff
@@ -476,6 +477,10 @@ where
             .context("Price feed not available")?
             .context("No quote available")?;
 
+        if latest_quote.is_older_than(1.minutes()) {
```
🤔 Perhaps a constant would make sense here. Also the error message could change depending on its value.

luckysori commented on daemon/src/routes_taker.rs:
```diff
@@ -100,9 +89,8 @@ pub async fn feed(
-                Ok(()) = rx_keep_alive.changed() => {
-                    let keep_alive = *rx_keep_alive.borrow();
-                    yield keep_alive.to_sse_event();
+                _ = heartbeat.tick() => {
+                    yield Heartbeat::new().to_sse_event();
```
👍 Hehe this is simpler 🤦

luckysori reviewed (approved):
LGTM!

thomaseizinger committed:
Use `next_order` to get orders from feed

Achieves the same thing but reusing an already existing function.

thomaseizinger committed:
Have `assert_next_state` wait until the given state

Previously, the macro would assert that the state of the CFD on the
next state change in the feed is a certain state. This can make the
tests brittle if there is another update in the feed inbetween where
the state is not yet the expected one.

Instead of asserting on the next feed update, we rewrite
`assert_next_state` as `wait_next_state` and spin in a loop until the
CFD satisfies a given predicate which in the case of the macro simply
performs an equality check on the state of the CFD.

This might potentially fix #1064.

thomaseizinger committed:
Increase timeout to 180 seconds

When running all tests locally, cargo by default executes as many
tests as you have cores. That is 8 for me. On 8 cores, only 8 threads
can meaningfully run concurrently. For each test that sets up a contract
tokio ends up spawning 1 additional thread for blocking CPU work per
party (maker & taker). We have more than 4 tests which means not all
of these threads can utilize the CPU 100% for their own work. This makes
the entire contract setup take quite a bit longer because the kernel
ends up scheduling all of these threads to get some work done.

In the meantime, the timeout value which we are changing here is ticking
inside the test because it is waiting for contract setup to finish. We
need to increase this to account for this potential delay.

da-kami commented on daemon/tests/happy_path.rs:
```diff
@@ -25,45 +26,25 @@ use std::time::Duration;
-        assert_eq!(maker_cfd.state, $state, "unexpected cfd state in the maker");
-    };
+/// Waits until the CFDs for both maker and taker are in the given state.
+macro_rules! wait_next_state {
```
I am not against this change, but I want to point out that 

> This can make the tests brittle if there is another update in the feed inbetween where the state is not yet the expected one.

this should not happen! What we call "brittle" here might actually be a flaw in the logic. With letting the mocked actors return very specific messages at specific points in time we can ensure that we receive the right message at the right time. I think we should always be able to ensure this. 
As @thomaseizinger pointed out we actually have e.g. quote updates that *do* send out projection updates independent from the tests. But I wonder if that *should* happen - shouldn't we be able to control this in the tests through the mocks? 

We *can* change the test setup to wait for the next state and ignore "other" state changes in between, but that makes the test less "restrictive" in testing the behavior.

thomaseizinger replied:
> this should not happen! What we call "brittle" here might actually be a flaw in the logic.

I don't think it is as clear cut to say that it shouldn't happen.

We cannot and IMO should not have control over ever single message that is sent in the system. That would require exposing way too many knobs and mocks. We have to accept that the actor system we are building are eventually consistent, meaning things will eventually be in the correct state if it works correctly.

> We _can_ change the test setup to wait for the next state and ignore "other" state changes in between, but that makes the test less "restrictive" in testing the behavior.

It would be fairly trivial to pass a list of states that it should go through before arriving at the final state. But outside of that, I don't think it is particularly likely for us to be in an unexpected state and then actually end up in the correct state in the end. And even if, is that really a bug or is that kind of self-healing functionality a feature? I don't think the user is gonna care much ...

da-kami commented on daemon/tests/harness/flow.rs:
```diff
@@ -8,7 +8,7 @@ use std::time::Duration;
 
 /// Waiting time for the time on the watch channel before returning error
-const NEXT_WAIT_TIME: Duration = Duration::from_secs(if cfg!(debug_assertions) { 120 } else { 30 });
+const NEXT_WAIT_TIME: Duration = Duration::from_secs(if cfg!(debug_assertions) { 180 } else { 30 });
```
I understand the reasoning in the commit message, but I am surprised that we only notice this *now*. 
Before we did event sourcing and concurrent actor setup I feel we did not run into these problems (neither locally nor on CI). Maybe it's a coincidence, but I still find it weird that we now set `180` when at some point `60` was enough to "not be brittle".

thomaseizinger replied:
This became more apparent by using spawn-blocking because I think before we were actually blocking the timer future so time did simply not progress!

da-kami commented on daemon/src/projection.rs:
```diff
@@ -56,7 +61,12 @@ pub struct Feeds {
-    pub fn new(db: sqlx::SqlitePool, _role: Role, network: Network) -> (Self, Feeds) {
+    pub fn new(
+        db: sqlx::SqlitePool,
+        _role: Role,
```
Why do we pass the role in here if we are not using it?

thomaseizinger replied:
Not sure, was already there as you can see. Do a follow-up refactor?

luckysori commented on daemon/tests/happy_path.rs:
```diff
@@ -25,45 +26,25 @@ use std::time::Duration;
-        );
+    };
+    ($id:expr, $maker:expr, $taker:expr, $state:expr) => {
+        wait_next_state!($id, $maker, $taker, $state, $state)
```
❓ Why does this branch of the macro not use `one_cfd_with_state` like above?

thomaseizinger replied:
This recurses. How should I use the filter-map function here?

luckysori replied:
Lol, please ignore me. I can't read code anymore.

luckysori commented on daemon/tests/harness/flow.rs:
```diff
@@ -62,15 +37,56 @@ pub async fn next<T>(rx: &mut watch::Receiver<T>) -> Result<T>
+            match filter_map(current) {
+                Some(val) => return anyhow::Ok(val),
+                None => {
+                    rx.changed().await?;
```
❓ Why don't we continue onto the next iteration of the loop and wait for `changed` there?

thomaseizinger replied:
So it turns out that the semantics of waiting for the next change are still useful for some tests where we don't use the macro.

luckysori replied:
I'm not getting you. Why are we awaiting _twice_ in a row for another change after not finding the `expected_state`. Wouldn't this code just ignore one state update?

thomaseizinger replied:
Ah you are right. That doesn't make any sense!

luckysori reviewed (approved):
Looks good again, just some questions about the first few new commits.

da-kami commented on daemon/src/bitmex_price_feed.rs:
```diff
@@ -212,4 +220,30 @@ mod tests {
+    fn quote_from_one_hour_ago_is_old() {
+        let quote = dummy_quote_at(OffsetDateTime::now_utc() - 1.hours());
+
+        let is_older = quote.is_older_than(1.minutes());
```
nit: The test would be more meaningful if it would test the boundaries. E.g.:

```
let quote = dummy_quote_at(OffsetDateTime::now_utc() - 2.minutes());

let is_older = quote.is_older_than(1.minutes());
```

might also want to cover more boundaries.

thomaseizinger replied:
Well, it is really a nitpick isn't it :D

The test makes sure we are doing the comparison correctly, i.e. older and not more recent. The fact that it uses one minute is pretty obvious from the impl so risk is low, hence I didn't bother :)

da-kami replied:
🤷‍♂️

da-kami reviewed (approved):
Thanks! few comments, nothing blocking

luckysori reviewed (approved):
Thanks for fixing that!

klochowicz commented on daemon/tests/harness/flow.rs:
```diff
@@ -62,15 +37,53 @@ pub async fn next<T>(rx: &mut watch::Receiver<T>) -> Result<T>
+/// framework.
+pub fn one_cfd_with_state(expected_state: CfdState) -> impl Fn(Vec<Cfd>) -> Option<Cfd> {
+    move |cfds: Vec<Cfd>| match cfds.as_slice() {
+        [one] if one.state == expected_state => Some(one.clone()),
+        [_one_that_doesnt_match_state] => None,
+        [] => None,
+        _more_than_one => panic!("More than one CFD in feed!"),
```
never seen this syntax before - what does it mean? are the `[]` destructuring the array of cfds?

thomaseizinger replied:
This is called sub-slice pattern matching: https://github.com/rust-lang/rfcs/blob/master/text/2359-subslice-pattern-syntax.md

klochowicz commented on daemon/tests/harness/mod.rs:
```diff
@@ -310,15 +325,23 @@ macro_rules! deliver_event {
+pub fn dummy_quote() -> Quote {
+    Quote {
+        timestamp: Timestamp::now(),
+        bid: Price::new(dec!(50_000)).expect("to not fail"),
+        ask: Price::new(dec!(50_000)).expect("to not fail"),
```
nit: would probably use the `dummy_price()` inside dummy quote, no need to have more magic numbers than necessary

klochowicz reviewed (approved):
nice, thanks for waiting for me to have a look ;)
